File: lib/src/platform/VirgilStreamDataSource.cxx
#include <virgil/service/stream/VirgilStreamDataSource.h>
using virgil::service::stream::VirgilStreamDataSource;

#include <virgil/VirgilByteArray.h>
using virgil::VirgilByteArray;

#include <algorithm>

static const size_t kChunkSizeMin = 32;

VirgilStreamDataSource::VirgilStreamDataSource(std::istream& in, size_t chunkSize)
        : in_(in), chunkSize_(std::max(chunkSize, kChunkSizeMin)) {
}

VirgilStreamDataSource::~VirgilStreamDataSource() throw() {
}

bool VirgilStreamDataSource::hasData() {
    return !in_.eof();
}

VirgilByteArray VirgilStreamDataSource::read() {
    size_t readBytesCount = 0;
    VirgilByteArray data;
    char byte;
    while (readBytesCount < chunkSize_ && in_.get(byte)) {
        data.push_back(byte);
    }
    return data;
}

File: wrappers/java/src/VirgilStreamDataSource.java
/**
 * Copyright (C) 2015-2016 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.virgilsecurity.crypto;

public class VirgilStreamDataSource extends VirgilDataSource implements java.io.Closeable {
    private java.io.InputStream stream;
    private int chunkSize;
    private static int CHUNK_SIZE_DEFAULT = 1024 * 1024;

    public VirgilStreamDataSource(java.io.InputStream stream) {
        this.stream = stream;
        this.chunkSize = CHUNK_SIZE_DEFAULT;
    }

    public VirgilStreamDataSource(java.io.InputStream stream, int chunkSize) {
        this.stream = stream;
        this.chunkSize = chunkSize;
    }

    @Override
    public boolean hasData() throws java.io.IOException {
        return this.stream.available() > 0;
    }

    @Override
    public byte[] read() throws java.io.IOException {
        final int bytesToReadNum = Math.min(this.stream.available(), this.chunkSize);
        byte[] result = new byte[bytesToReadNum];
        this.stream.read(result, 0, bytesToReadNum);
        return result;
    }

    @Override
    public void close() throws java.io.IOException {
        this.stream.close();
        this.delete();
    }
}
